sw/device/lib/dif/dif_hmac.c

```diff
@@ -227,48 +200,115 @@ dif_hmac_result_t dif_hmac_irq_force(const dif_hmac_t *hmac,
+  switch (config.message_endianness) {
+    case kDifHmacEndiannessBig:
+      new_config =
+          bitfield_bit32_write(new_config, HMAC_CFG_ENDIAN_SWAP, false);
+      break;
+    case kDifHmacEndiannessLittle:
+      new_config = bitfield_bit32_write(new_config, HMAC_CFG_ENDIAN_SWAP, true);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  // Set the byte-order of the digest.
+  switch (config.digest_endianness) {
+    case kDifHmacEndiannessBig:
+      new_config =
+          bitfield_bit32_write(new_config, HMAC_CFG_DIGEST_SWAP, false);
+      break;
+    case kDifHmacEndiannessLittle:
+      new_config = bitfield_bit32_write(new_config, HMAC_CFG_DIGEST_SWAP, true);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  *device_config = new_config;
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_mode_hmac_start(
+    const dif_hmac_t *hmac, const uint8_t *key,
+    const dif_hmac_transaction_config_t config) {
   if (hmac == NULL || key == NULL) {
     return kDifHmacBadArg;
   }
 
-  // Disable SHA256 while configuring and to clear the digest.
-  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                    HMAC_CFG_SHA_EN);
-
   // Set the HMAC key.
   // TODO Static assert register layout.
   mmio_region_memcpy_to_mmio32(hmac->base_addr, HMAC_KEY_0_REG_OFFSET, key,
                                HMAC_PARAM_NUMWORDS * sizeof(uint32_t));
 
+  // Read current CFG register value.
+  uint32_t device_config =
+      mmio_region_read32(hmac->base_addr, HMAC_CFG_REG_OFFSET);
+
+  // Set the byte-order of the input message and the digest.
+  dif_hmac_result_t update_result =
+      dif_hmac_calculate_device_config_value(&device_config, config);
+  if (update_result != kDifHmacOk) {
+    return update_result;
+  }
+
   // Set HMAC to process in HMAC mode (not SHA256-only mode).
-  mmio_region_nonatomic_set_mask32(
-      hmac->base_addr, HMAC_CFG_REG_OFFSET,
-      (1 << HMAC_CFG_SHA_EN) | (1 << HMAC_CFG_HMAC_EN), 0);
+  device_config = bitfield_bit32_write(device_config, HMAC_CFG_SHA_EN, true);
+  device_config = bitfield_bit32_write(device_config, HMAC_CFG_HMAC_EN, true);
+  mmio_region_write32(hmac->base_addr, HMAC_CFG_REG_OFFSET, device_config);
 
   // Begin HMAC operation.
   mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
                                   HMAC_CMD_HASH_START);
   return kDifHmacOk;
 }
 
-dif_hmac_result_t dif_hmac_mode_sha256_start(const dif_hmac_t *hmac) {
+dif_hmac_result_t dif_hmac_mode_sha256_start(
+    const dif_hmac_t *hmac, const dif_hmac_transaction_config_t config) {
   if (hmac == NULL) {
     return kDifHmacBadArg;
   }
 
-  // Disable SHA256 while configuring and to clear the digest.
-  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                    HMAC_CFG_SHA_EN);
+  // TODO: Clear HMAC Key? Write Zeroed HMAC Key?
```

Comment:
What's the right thing to do here?

Reply:
Paging @jon-flatley

Reply:
I'm not sure, the key is ignored in SHA256-only operation and is clobbered by the new key in HMAC mode. From purely a correctness perspective I don't think there's any reason to clear the key here (or anywhere?) I think this more so matters from a security perspective. The right thing to do here is expose `hmac.WIPE_SECRET` imo, I'm not sure why I didn't include that in the API. It doesn't hurt to also clear the key here, especially since an existing key can't be reused by the DIF.

Reply:
The DIF is a good place to enforce good security practice. If we want to make sure the key is cleared before using it in this mode, doing so in the DIF is a-ok, as long as we do this in the way the hardware is expecting.

Btw, you do cover WIPE_SECRET - with `dif_hmac_wipe_secret`.

Maybe a better question is this: will the interface for sharing the key change as this peripheral is hardened? I think this happened with AES, it could happen with HMAC too, so maybe we should wait until we know what the hardening story is here, and leave the TODO in the meantime.